### ISSUE TITLE
Date Encoder crashes on encoding Mon 00:00 

### ISSUE THREAD

**Issue by shahardror** [htm-community/htm.core#578]
In the C++ code, if I understand correctly, right now the scalar encoder clips all the inputs to [min,max] and does not implement a periodic encoder. In lines 152-153 there is the following "clipping":
```
      input = std::max(input, parameters.minimum);
      input = std::min(input, parameters.maximum);
```
which only makes sure that input<=maximum. If input<=min, then input does not change.

This confuses me because I do not understand how the hotgym example is working, since there are certain times that need to be encoded that are out of bounds. An example is any time on a Monday before noon which causes dayOfWeek<=0 where [min,max]=[0,7].
I would be happy for any clarification on my misunderstanding, because clearly I am missing something.

Thank you!

**Comment by breznak** [comment]
> which only makes sure that input<=maximum. If input<=min, then input does not change.

no, it checks (rounds) to both lower,upper bounds. 

this is for the lower bound 

> input = std::max(input, parameters.minimum);

>  An example is any time on a Monday before noon which causes dayOfWeek<=0 where [min,max]=[0,7].

according to what you've posted, this would become `0`

**Comment by shahardror** [comment]
Hi,

I am having trouble understanding how to work with the Date Encoder (date.py) and would be happy for more detail. For example, I am running the following bit of code and have a few questions regarding the output.

1. I am getting encodings of different length that were initialized with the same parameters. Can you please explain why it happens? Because I understood that the Date Encoder output is the concatenation of several sub-encodings.
2. If I change the date to `datetime.strptime("22/07/19 8:00", "%d/%m/%y %H:%M")` the encoder `Encoder` fails to encode `date` which is unclear to me why it happens. The concatenated encoder does not experience any issues.

Thank you!

```
from htm.encoders.date import DateEncoder
            
dayOfWeek = 3
timeOfDay = 10
Encoder = DateEncoder(dayOfWeek, timeOfDay)

timeOfDayEncoder  = DateEncoder(timeOfDay)
dayOfWeekEncoder  = DateEncoder(dayOfWeek)
encodingWidth     = (timeOfDayEncoder.size + dayOfWeekEncoder.size)

date = datetime.strptime("23/07/19 8:00", "%d/%m/%y %H:%M")

Bits  = Encoder.encode(date)

timeOfDayBits     = timeOfDayEncoder.encode(date)
dayOfWeekBits     = dayOfWeekEncoder.encode(date)
# Concatenate all these encodings into one large encoding for Spatial Pooling.
encoding = SDR( encodingWidth ).concatenate([timeOfDayBits, dayOfWeekBits])

print( date )
print( encodingWidth )
print( encoding.dense )

print( Encoder.size )
print( Bits.dense )
```

**Comment by breznak** [comment]
ok, this is a running example. 

```
from datetime import datetime
from htm.encoders.date import DateEncoder
from htm.bindings.sdr import SDR
Encoder = DateEncoder(dayOfWeek=3, timeOfDay=10)
timeOfDayEncoder  = DateEncoder(timeOfDay=10)
dayOfWeekEncoder  = DateEncoder(dayOfWeek=3)
encodingWidth     = (timeOfDayEncoder.size + dayOfWeekEncoder.size)
date = datetime.strptime("23/07/19 8:00", "%d/%m/%y %H:%M")
Bits  = Encoder.encode(date)
timeOfDayBits     = timeOfDayEncoder.encode(date)
dayOfWeekBits     = dayOfWeekEncoder.encode(date)
# Concatenate all these encodings into one large encoding for Spatial Pooling.
encoding = SDR( encodingWidth ).concatenate([timeOfDayBits, dayOfWeekBits])
print( date )
print( encodingWidth )
print( encoding.dense )
print( Encoder.size )
print( Bits.dense )
assert Encoder.size == timeOfDayEncoder.size + dayOfWeekEncoder.size, "sizes mismatch! "+str(Encoder.size)+" vs "+str(timeOfDayEncoder.size)+" + "+str(dayOfWeekEncoder.size)
assert Bits == encoding
```

the problem was in how you initialize objects in python, this is not doing what you've expected

> dayOfWeek = 3
timeOfDay = 10
Encoder = DateEncoder(dayOfWeek, timeOfDay)

see in the code example the correct instatiation. 

- [x] now the sizes match
- [ ] I have left the encodings mismatch, you can see the difference is in order how you concat the 2 internal representations. 

TL;DR: you should use `Encoder = DateEncoder(dayOfWeek=3, timeOfDay=10)`

**Comment by shahardror** [comment]
Thank you for the explanation and the patience!

**Comment by breznak** [comment]
no worries, reopen again if you have a question or a problem, cheers

**Comment by shahardror** [comment]
If you change the date to Monday morning, specifically `datetime.strptime("22/07/19 8:00", "%d/%m/%y %H:%M")` the `dayOfWeek `encoder runs into the following error:
`RuntimeError: CHECK FAILED: "input >= parameters.minimum && input <= parameters.maximum" Input must be within range [minimum, maximum]!`
This is unclear to me since the Scalar Encoder is meant to clip the input. Any insight to why it could occur would be helpful.
Thanks again!

**Comment by breznak** [comment]
> f you change the date to Monday morning, specifically datetime.strptime("22/07/19 8:00", "%d/%m/%y %H:%M") the dayOfWeek encoder runs into the following error:

indeed, good call! 

I'm starting a debug branch for this, this is the details from cpp ScalarEnc

> RR:  CHECK FAILED: "input >= parameters.minimum && input <= parameters.maximum" Input must be within range [minimum, maximum]! -0.166667 vs [0 , 7 ] [/mnt/store/devel/HTM/htm-community/nupic.cpp/src/htm/encoders/ScalarEncoder.cpp line 161]
Traceback (most recent call last):
File "dateSizeCheck.py", line 14, in <module>
Bits  = Encoder.encode(date)
File "/mnt/store/devel/HTM/pyenv3/lib/python3.6/site-packages/htm.core-2.0.0-py3.6-linux-x86_64.egg/htm/encoders/date.py", line 260, in encode
sdrs.append( self.dayOfWeekEncoder.encode(dayOfWeek) )
RuntimeError: CHECK FAILED: "input >= parameters.minimum && input <= parameters.maximum" Input must be within range [minimum, maximum]! -0.166667 vs [0 , 7 ]

@shahardror 
- [ ] can you add this test-case to the DateTime enc py tests, so we can trigger this? (just add the test with the minimal example you've provided above) 
- [ ] DateTime is using py's `scalar_encoder`, which is a wrapper for bindings::ScalarEncoder, but forces `clipInput = False`
  - I'm testing if using correctly the one from bindings fixes the issue?
- [ ] the scalar_encoder should probably moved to "only" examples/

**Comment by breznak** [comment]
@shahardror will you have time to finish the PR, please?
